arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -265,6 +266,113 @@ EvalResult Prim_Dict(Machine& ctx, VPackSlice const params, VPackBuilder& result
   return {};
 }
 
+EvalResult Prim_DictKeys(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (params.length() != 1) {
+    return EvalError("expected exactly one parameter");
+  }
+
+  auto obj = params.at(0);
+  if (!obj.isObject()) {
+    return EvalError("expected object, found: " + obj.toJson());
+  }
+
+  result.openArray();
+  for (VPackObjectIterator iter(params.at(0)); iter.valid(); iter++) {
+    result.add(iter.key());
+  }
+  result.close();
+
+  return {};
+}
+
+std::list<std::string> createObjectPaths(velocypack::Slice object,
+                                         std::list<std::string> currentPath) {
+  for (VPackObjectIterator iter(object); iter.valid(); iter++) {
+    if (iter.value().isObject()) {
+      // recursive through all available keys
+      currentPath.emplace_back(iter.key().toString());
+      return createObjectPaths(iter.value(), currentPath);
+    } else {
+      // reached end
+      currentPath.emplace_back(iter.key().toString());
+    }
+  }
+  return currentPath;
+}
+
+void printPath(std::list<std::string>& path) {
+  // TODO: can be removed - internal debugging method
+  std::cout << "Printing current path: " << std::endl;
+  std::cout << " [ ";
+  for (auto const& pathElement : path) {
+    std::cout << " " << pathElement << " ";
+  }
+  std::cout << " ] " << std::endl;
+}
+
+void createPaths(std::list<std::list<std::string>>& finalPaths,
+                 velocypack::Slice object, std::list<std::string>& currentPath) {
+  for (VPackObjectIterator iter(object); iter.valid(); iter++) {
+    std::string currentKey = iter.key().toString();
+    VPackSlice currentValue = iter.value();
+
+    if (currentValue.isObject()) {
+      // path not done yet
+      currentPath.emplace_back(currentKey);
+
+      std::list<std::string> currentTmpPath(currentPath);
+      finalPaths.emplace_back(currentTmpPath);
+      createPaths(finalPaths, currentValue, currentPath);
+    } else {
+      // path is done
+      std::list<std::string> currentTmpPath(currentPath);
+      currentTmpPath.emplace_back(currentKey);
+      finalPaths.emplace_back(currentTmpPath);
+    }
+
+    if (iter.isLast()) {
+      // if end of path reached, remove last visited member
+      if (currentPath.size() > 0) {
+        currentPath.pop_back();
+      }
+    }
+  }
+}
+
```

Comment:
Is this tested at the moment?

Reply:
It is. That is being used within the dict-directory primitive. And that one is used in the JS test. But yes, we should add standalone primitive test as well!